### ISSUE TITLE
protobuf packaging doesnt retain directory structure in Android aar libraries

### ISSUE THREAD

**Issue by nydrani** [google/protobuf-gradle-plugin#435]
Generating a protobuf library on android doesn't retain the directory structure from the base "src/main/proto" folder.

e.g. A folder structure of:
```
src/main/proto
├── a.proto
├── firstfolder
│   └── b.proto
└── secondfolder
    └── c.proto
```

will package all corresponding .protobuf files into the root directory of the packaged .aar file instead of in the corresponding folders (firstfolder and secondfolder).

This means that applications consuming the protobuf library and trying to import the `b.proto` file using
`import "firstfolder/b.proto"` will fail.

I assume the inclusion of proto files is here: https://github.com/google/protobuf-gradle-plugin/blob/master/src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy#L273 or https://github.com/google/protobuf-gradle-plugin/blob/master/src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy#L293. But I'm not quite sure why the directory structure is not retained.

**Comment by voidzcy** [comment]
I am able to reproduce the bug locally. There are two issues involved here:

The way proto files being packaged does not preserve the directory structure.
https://github.com/google/protobuf-gradle-plugin/blob/7f964837a825af2a3e9c32d61889b66aa98b22a1/src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy#L292-L293

In order to preserve the original directory hierarchy, this should be achieved by the variant's `ProcessResourceTask` (which is a Copy task that preserves the directory hierarchy). This should be similar to what's being done for Java projects:
```groovy
Task processResourcesTask = variant.getProcessJavaResources()
processResourcesTask.from(generateProtoTask.sourceFiles) {
    include '**/*.proto'
}

variant.sourceSets.each {
    Task extractTask = setupExtractProtosTask(generateProtoTask, it.name)
    processResourcesTask.dependsOn(extractTask)
}
```

The more critical problem is in the proto extraction side. The plugin is not able to extract proto files from `.aar`:
https://github.com/google/protobuf-gradle-plugin/blob/7f964837a825af2a3e9c32d61889b66aa98b22a1/src/main/groovy/com/google/protobuf/gradle/ProtobufExtract.groovy#L121-L144


It turns out the test added by #414 does not test the real thing. [Depending on a subproject](https://github.com/google/protobuf-gradle-plugin/blob/728b686b61ddf0e5e7bc72eb829ae7f11a05e67a/testProjectAndroidBase/build_base.gradle#L113) does not go through the packaging process, dependent classes and resources are pulled directly from the subproject. Also, the test is weak as it only covers the consumption of generated classes. The application's protos are not consuming protos in the library project, which is the more important behavior we want to see.

Some other existing tests (e.g., the [library-application test](https://github.com/google/protobuf-gradle-plugin/blob/7f964837a825af2a3e9c32d61889b66aa98b22a1/testProjectDependentApp/build.gradle#L18) for Java projects) may also not test the desired behavior, and the test itself is also weak.

I am probably going to revert #414. I will fix this bug and use an `aar` file as as an included library for testing, similar to [what we are doing for Java](https://github.com/google/protobuf-gradle-plugin/tree/v0.8.13/testProjectBase/lib).

**Comment by robcos** [comment]
I'm piggybacking on this issue as I think what I see is related.
With 0.8.13 I can _normally_ see the .protos and, as said in this bug, they are stripped from the path, which is wrong.

However, sometimes (and it seems to be consistent), when I build a full app, the res.jar with the protos is not generated at all. I can't share full logs but maybe the attached screenshot can help understand what is going on.

Build A is a good one, Build B is a bad one. During Build B, there is no res.jar in the build folder.
As you can see the :libs:proto:processDebugJavaRes has different value properties. 
The next step _bundleLibResDebug_ has no value for the resources and resourcesAsJars in build B:

<img width="815" alt="Screenshot 2020-10-07 at 15 38 12" src="https://user-images.githubusercontent.com/64998/95345869-2225c180-08b3-11eb-81a6-5bb344f4dfb3.png">

**Comment by robcos** [comment]
Is there anything I can do to help this bug moving forward? It is preventing my company (and anyone else using gradle I guess!) to really enjoy the power of protos. You have no idea of the hacks we have to go through otherwise :-)

**Comment by nydrani** [comment]
> 
> 
> Is there anything I can do to help this bug moving forward? It is preventing my company (and anyone else using gradle I guess!) to really enjoy the power of protos. You have no idea of the hacks we have to go through otherwise :-)

The workaround I used was to build a java library (.jar) instead of an android library (.aar).

**Comment by robcos** [comment]
I'm actually seeing this issue with jars. Am I doing something wrong?

**Comment by voidzcy** [comment]
> I'm actually seeing this issue with jars. Am I doing something wrong?

Jars shouldn't have problems. Can you describe what you see or provide a minimal reproducible example?

**Comment by voidzcy** [comment]
The fix is in 0.8.14 release now.
